File: 027_LargestRectangleInHistogram.cpp
// Topic: Largest Rectangle in Histogram
// Description: Find largest rectangular area in histogram bars.
// Approach: Monotonic stack of indices to compute nearest smaller to left/right in O(n).

#include <bits/stdc++.h>
using namespace std;

class Solution
{
   public:
    int largestRectangleArea(const vector<int>& heights)
    {
        int n = heights.size();
        stack<int> st;
        int best = 0;
        for (int i = 0; i <= n; ++i)
        {
            int h = (i == n ? 0 : heights[i]);
            while (!st.empty() && h < heights[st.top()])
            {
                int height = heights[st.top()];
                st.pop();
                int left = st.empty() ? -1 : st.top();
                best = max(best, height * (i - left - 1));
            }
            st.push(i);
        }
        return best;
    }
};

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    vector<vector<int>> tests = {{2, 1, 5, 6, 2, 3}, {2, 4}};
    for (size_t i = 0; i < tests.size(); ++i)
    {
        cout << "Test " << (i + 1) << ": ";
        cout << Solution().largestRectangleArea(tests[i]) << "\n";
    }
}

/*
Time Complexity: O(n)
Space Complexity: O(n)
*/

File: 027_LargestRectangleInHistogram.py
# Topic: Largest Rectangle in Histogram
# Description: Find largest rectangular area in histogram bars.
# Approach: Monotonic stack of indices to compute nearest smaller to left/right in O(n).

from typing import List


class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        n = len(heights)
        stack = []
        best = 0
        for i in range(n + 1):
            h = heights[i] if i < n else 0
            while stack and h < heights[stack[-1]]:
                height = heights[stack.pop()]
                left = stack[-1] if stack else -1
                best = max(best, height * (i - left - 1))
            stack.append(i)
        return best


if __name__ == "__main__":
    tests = [[2, 1, 5, 6, 2, 3], [2, 4]]
    sol = Solution()
    for i, heights in enumerate(tests):
        result = sol.largestRectangleArea(heights)
        print(f"Test {i + 1}: {result}")

"""
Time Complexity: O(n)
Space Complexity: O(n)
"""
